### PR TITLE
chore(pipelined): datapath refactor inout

### LINKED ISSUE #12072: datapath: Refactor in-out controller into three separate controllers
Inout controller has became very complicated due to managing three separate tables. it would be better to separate out it into separate APPs.
This task also involve separating unit tests for in-out controller.

verification:
1. run pipelineD unit tests
2. run integ tests.

### PR DESCRIPTION
**Pull request by nstng**
## Summary
See https://github.com/magma/magma/issues/12072.

Done in pairing with @vktng 

This PR is best reviewed commit by commit
1. Add ingress, middle and egress controller and super class mandatory_service. This is basically copying logic from inout.py to the three (+1 super class) new files. Note that for read- and reviewability code was just copied without any refactoring.
2. Switch sudo unit tests to using the new controllers
3. Make rpc servicer using new controllers.
4. Remove inout controller and use constants from new controllers.
5. Adapt documentation.
6. Reflect changes here in Bazel.

Open questions:
* #12072 states that unit tests should be split up as well - is it better to test the three controllers in context?
  * If this is still requested, then we suggest to do it in a follow up PR.

## Test Plan

* CI
* integ tests on fork: https://github.com/nstng/magma/actions/runs/2203084202

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/app/mandatory_service.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/mandatory_service.py]: Hm I'm not sure why we're adding unnecessary inheritance here, whats the purpose of MandatoryController class?

Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/mandatory_service.py]: I think we shouldn't introduce this

**Review by themarwhal — APPROVED** [review]
doc changes seem reasonable, I'll let the pipelined codeowners validate the correctness :D Also, out of scope for this PR but is the diagram better handled with mermaid? (https://mermaid-js.github.io/mermaid/#/)

File: lte/gateway/python/magma/pipelined/app/mandatory_service.py
Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/mandatory_service.py]: The reason is the common logic used in the three new controllers - would you prefer to handle this in a util class? (at least the ryu annotated functions would need to be duplicated in ingress, middle and egress apps).

File: lte/gateway/python/magma/pipelined/app/mandatory_service.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/mandatory_service.py]: f.e this seems to be only used in middle

File: lte/gateway/python/magma/pipelined/app/mandatory_service.py
Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/mandatory_service.py]: Yeah we already have a MagmaController class, lets either put stuff there(only if its applicable to all controllers) or duplicate the code

Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/mandatory_service.py]: Hi Nick, MagmaController is super class of a lot of apps so we should not move the common logic for ingress, middle and egress there. We will remove the new abstract controller and inline the code.

> f.e this seems to be only used in middle

_get_vlan_egress_flow_msgs is used in middle and egress and please note that this does not belong to the class MandatoryController - this is the same patter as in inout.py before. We will move it to a util file/module.

**Comment by nstng** [comment]
FYI for reviewers: added an additional commit for Bazel adaptions. Reason: the pipelined bazelification was merged after this PR was created and the changes here need to be reflected in Bazel.

File: lte/gateway/python/magma/pipelined/app/mandatory_service.py
Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/mandatory_service.py]: pushed - running integ test on fork to be sure
https://github.com/nstng/magma/actions/runs/2203084202

**Review by LKreutzer — APPROVED** [review]

File: lte/gateway/python/magma/pipelined/service_manager.py
Comment (sebathomas) [inline lte/gateway/python/magma/pipelined/service_manager.py]: I think `self._tables_by_app` already contains these three names?

**Review by sebathomas — APPROVED** [review]
Looking good, just a minor point.

File: lte/gateway/python/magma/pipelined/app/vlan_egress_flow_msgs.py
Comment (sebathomas) [inline lte/gateway/python/magma/pipelined/app/vlan_egress_flow_msgs.py]:
```suggestion
def get_vlan_egress_flow_msgs(
```
This is used in another file, so it's not private.

File: lte/gateway/python/magma/pipelined/service_manager.py
Comment (nstng) [reply lte/gateway/python/magma/pipelined/service_manager.py]: addressed with last push

File: lte/gateway/python/magma/pipelined/app/vlan_egress_flow_msgs.py
Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/vlan_egress_flow_msgs.py]: addressed with last push

**Review by Neudrino — APPROVED** [review]
Only reviewed the "docs" change, which LGTM.

**Comment by Neudrino** [comment]
Afaiks there would be review needed again from @koolzz or @pshelar.
Could you review again, please, as this is blocking the next PR https://github.com/magma/magma/pull/12566 already.

File: lte/gateway/python/magma/pipelined/app/vlan_egress_flow_msgs.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/vlan_egress_flow_msgs.py]: Move this file up one directory and rename it to vlan_utils.py

**Review by koolzz — APPROVED** [review]
I have one nit please fix before merging. Otherwise looks really clean, thanks for working on it!
@pshelar give this a quick glance through maybe

File: lte/gateway/python/magma/pipelined/app/vlan_egress_flow_msgs.py
Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/vlan_egress_flow_msgs.py]: :+1: pushed

File: lte/gateway/python/magma/pipelined/vlan_utils.py
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/vlan_utils.py:91]: did you ran non-nat datapath tests from S1AP integ tests with this PR?

Comment (nstng) [reply lte/gateway/python/magma/pipelined/vlan_utils.py:91]: @pshelar, we run all tests in the "lte integ tests" workflow. Last run on fork: https://github.com/nstng/magma/runs/6115971885?check_suite_focus=true
Edit latest run: https://github.com/nstng/magma/actions/runs/2252702388

Comment (pshelar) [reply lte/gateway/python/magma/pipelined/vlan_utils.py:91]: I dont see `test_attach_detach_non_nat_dp_ul_tcp.py` in the test log, can you check it.

Comment (nstng) [reply lte/gateway/python/magma/pipelined/vlan_utils.py:91]: `test_attach_detach_non_nat_dp_ul_tcp.py` is in the `NON_SANITY_TESTS` list. It looks like the `NON_SANITY_TESTS` are not executed with the LTE integ-tests workflow. I do not see them executed anywhere automatically. Do you happen to know if this is intentionally?
I will execute them manually - but this will take some time.

Comment (pshelar) [reply lte/gateway/python/magma/pipelined/vlan_utils.py:91]: Lets move this test to 'PRECOMMIT_TESTS' set.
cc: @ssanadhya

Comment (nstng) [reply lte/gateway/python/magma/pipelined/vlan_utils.py:91]: @pshelar executed the tests locally - `test_attach_detach_non_nat_dp_ul_tcp.py` was successful.

In general the tests seem to be very flaky (on a fresh vm setup). I still do not have all tests green. Do you have an opinion if one of the following tests might be relevant for the change?:
```
test_s1_handover_ping_pong.py
test_activate_deactivate_multiple_dedicated.py
test_sctp_shutdown_while_mme_is_stopped.py
test_secondary_pdn_with_dedicated_bearer_multiple_services_restart.py
test_3495_timer_for_dedicated_bearer_with_mme_restart.py
test_no_attach_complete_with_mme_restart.py
```

Comment (pshelar) [reply lte/gateway/python/magma/pipelined/vlan_utils.py:91]: these are not related datapath, so we can ignore it.

**Comment by prabinakpattnaik** [comment]
Looks good to me for these changes.

File: lte/gateway/python/magma/pipelined/app/middle.py
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/middle.py:139]: function, _install_default_flows() is duplicated in ingress, middle and egress, can we move it to base calss.

File: lte/gateway/python/magma/pipelined/app/middle.py
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/middle.py:148]: same as above, this function can be moved to base class.

File: lte/gateway/python/magma/pipelined/app/middle.py
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/middle.py:160]: same for this function.

**Review by pshelar — CHANGES_REQUESTED** [review]

File: lte/gateway/python/magma/pipelined/app/middle.py
Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/middle.py:139]: @pshelar: base.py is the super class of other controllers too. And those other controllers also have `_install_default_flows()` - but these are diverging implementations. Do we really want to implement `_install_default_flows()` of ingress, middle and egress in base.py and let all other controllers override this function?
Another approach would be to add an inheritance layer between base.py and ingress, middle and egress and put common logic there - this was already discussed with @koolzz above - this has the downside of adding inheritance complexity to the model.

File: lte/gateway/python/magma/pipelined/app/middle.py
Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/middle.py:148]: See answer above - additionally this function is not used by a lot of controllers that extend base.by.

File: lte/gateway/python/magma/pipelined/app/middle.py
Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/middle.py:160]: See answers above.

File: lte/gateway/python/magma/pipelined/app/middle.py
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/app/middle.py:139]: ok, thats fine.

**Review by pshelar — APPROVED** [review]
Thanks for working on this.

**Pull request merged**
